Release notes for 2022.9.0 release. (#2267)
uxlfoundation/oneDPL: documentation/release_notes.rst

Known Issues and Limitations
----------------------------
New in This Release
^^^^^^^^^^^^^^^^^^^

Existing Issues
^^^^^^^^^^^^^^^
See oneDPL Guide for other `restrictions and known limitations`_.

- Incorrect results may be observed when calling ``sort`` with a device policy on Intel® Arc™ graphics 140V with data
  sizes of 4-8 million elements.
- ``histogram`` algorithm requires the output value type to be an integral type no larger than four bytes
  when used with a device policy on hardware that does not support 64-bit atomic operations.
- Compilation issues may be encountered when passing zip iterators to ``exclusive_scan_by_segment`` on Windows. 
- For ``transform_exclusive_scan`` and ``exclusive_scan`` to run in-place (that is, with the same data
  used for both input and destination) and with an execution policy of ``unseq`` or ``par_unseq``, 
  it is required that the provided input and destination iterators are equality comparable.
  Furthermore, the equality comparison of the input and destination iterator must evaluate to true.
  If these conditions are not met, the result of these algorithm calls is undefined.

> - Incorrect results may be produced by ``exclusive_scan``, ``inclusive_scan``, ``transform_exclusive_scan``,
>   ``transform_inclusive_scan``, ``exclusive_scan_by_segment``, ``inclusive_scan_by_segment``, ``reduce_by_segment``
>   with ``unseq`` or ``par_unseq`` policy when compiled by Intel® oneAPI DPC++/C++ Compiler
>   with ``-fiopenmp``, ``-fiopenmp-simd``, ``-qopenmp``, ``-qopenmp-simd`` options on Linux.
>   To avoid the issue, pass ``-fopenmp`` or ``-fopenmp-simd`` option instead.

Reviewer: ```suggestion
- Incorrect results may be produced by ``exclusive_scan``, ``inclusive_scan``, ``transform_exclusive_scan``,
  ``transform_inclusive_scan``, ``exclusive_scan_by_segment``, ``inclusive_scan_by_segment``, ``reduce_by_segment``
  with ``unseq`` or ``par_unseq`` policy when compiled by Intel® oneAPI DPC++/C++ Compiler 2024.1 or earlier
  with ``-fiopenmp``, ``-fiopenmp-simd``, ``-qopenmp``, ``-qopenmp-simd`` options on Linux.
  To avoid the issue, pass ``-fopenmp`` or ``-fopenmp-simd`` option instead.
```

I've checked that this issue is fixed in icpx 2024.2 and newer.

Author: Done.

and provides high-productivity APIs aimed to minimize programming efforts of C++ developers
creating efficient heterogeneous applications.

New in 2022.9.0
===============

New Features
------------

> - Added parallel range algorithms in ``namespace oneapi::dpl::ranges``: ``fill``, ``move``, ``replace``, ``replace_if``,
> ``remove``, ``remove_if``, ``mismatch``, ``minmax_element``, ``min``, ``max``, ``find_first_of``, ``find_end``, ``is_sorted_until``

Reviewer: ```suggestion
- Added parallel range algorithms in ``namespace oneapi::dpl::ranges``: ``fill``, ``move``, ``replace``, ``replace_if``,
  ``remove``, ``remove_if``, ``mismatch``, ``minmax_element``, ``min``, ``max``, ``find_first_of``, ``find_end``, ``is_sorted_until``.
```

Author: is the added whitespace at the beginning of the line important?

Reviewer: Yes. Otherwise the second line is treated as a an individual block (not a part of a bullet).

> - Improved performance of set operation algorithms (e.g., ``set_union``, ``set_difference``, etc.)

Reviewer: ```suggestion
- Improved performance of set operation algorithms: ``set_union``, ``set_difference``, ``set_intersection``, ``set_symmetric_difference``. 
```

There are not many set algorithms. Let's specify them all.

Author: There are also set_union_by_key, set_intersection_by_key, set_symmetric_differency_by_key, and set_difference_by_key.  This is too long in my opinion.

Reviewer 2: the by_key algorithms are in the SYCLomatic helper headers, not oneDPL itself, so I don't think we need to include those here.  I dont have a strong opinion about whether to list everything out though.

Fixed Issues
------------
- Eliminated runtime exceptions that were encountered when compiling code that uses device policies with the open source

> oneAPI DPC++ Compiler (clang++ driver) without specifying an optimization flag.

Reviewer: One may think that the whole device backend was broken with that compiler with device policies. Let's make the note more specific. @mmichel11, could you make a suggestion?

Reviewer 3: This was affecting `scan` based algorithms.

I would phrase it, "when using device policies with ``inclusive_scan``, ``copy_if``, ``partition``, ``unique``, ``reduce_by_segment``, and related algorithms with the open source..."

Author: @mmichel11 what are the "and related algorithms"?

Reviewer 3: By related algorithms, I mean the other scan-based algorithms in the same family: things like `exclusive_scan`, `remove_if`, `partition_copy`, `unique_copy`, etc.

Author: I've made the suggested change.

> - Eliminated compile time warnings throughout the library.

Reviewer: ```suggestion
- Eliminated multiple compile time warnings throughout the library.
```

Still, some warning remain. Hence, I'm suggesting adding "multiple".

Author: Done.

Known Issues and Limitations
----------------------------
New in This Release
^^^^^^^^^^^^^^^^^^^

Existing Issues
^^^^^^^^^^^^^^^
See oneDPL Guide for other `restrictions and known limitations`_.

> - Incorrect results may be observed when calling ``sort`` with a device policy on Intel® Arc™ graphics 140V with data
>   sizes of 4-8 million elements.

Reviewer: ```suggestion
- ``sort``, ``stable_sort``, ``sort_by_key`` and ``stable_sort_by_key`` algorithms with a device policy are not supported
  on Windows with Intel® Arc™ graphics 140V yet.
```

Let's better say that it is not supported yet - the HW is relatively new.

Reviewer: Ah, you are right. Let's keep as is here. I hope to investigate and fix it by the next release.

- ``sort``, ``stable_sort``, ``sort_by_key`` and ``stable_sort_by_key`` algorithms fail to compile
  when using Clang 17 and earlier versions, as well as compilers based on these versions,

>   such as Intel® oneAPI DPC++/C++ Compiler 2023.2.0.

Reviewer: That's not relevant (introduced and then fixed between 2022.8 and 2022.9 releases)

Reviewer: ```suggestion
```

Author: Removed.

- ``histogram`` algorithm requires the output value type to be an integral type no larger than four bytes

>   when used with an FPGA policy.

Reviewer: Is it related to FPGA specifically or to, for example, any hardware that does not support 64-bit atomic operations? @danhoeflinger, could you clarify?

Reviewer 2: This would be any hardware which does not support 64-bit atomic operations.

Reviewer 2: Looks good.

> - ``histogram`` may provide incorrect results with device policies in a program built with ``-O0`` option.

Reviewer: Does it happen with specific hardware or their types (GPU/CPU)? Is there the same issue in Windows (e.g. with /O2 option)?

Reviewer 2: Good question.  This was a GPU specific issue which has since been fixed by an update in the drivers.  We can remove this line.

Reviewer 2: It looks like (rolling) version 2506.18. I'll see if I can figure out the LTS version.

Author: Info on the release dates of the rolling and LTS drivers is available at https://dgpu-docs.intel.com/releases/releases.html

Reviewer 2: I believe it is not specific to OS or GPU platform. This issue was showing up on windows and with integrated graphics as well as Intel Data Center GPU Max 1550. It also seems invariant of compiler version.  The driver seems to be the culprit.

Reviewer 2: Looks good.  resolving.

> - Compilation issues may be encountered when passing zip iterators to ``exclusive_scan_by_segment`` on Windows. 

Reviewer: Is there any workaround? @mmichel11 could you clarify?

Reviewer 3: I will take a look to see if we are still affected by this. This issue may be stale by now (it was introduced in 2022.3.0). If I recall, it was related to device copyability of some of our internal functors which should be fixed.

Reviewer 3: I verified we are no longer affected by this issue on Windows.

@timmiesmith Can this note be removed as a known limitation?

Author: @mmichel11 was the device copyability issue fixed since the last release?  If so we need to list this as a fixed issue instead of just removing it.

Reviewer 3: This was fixed as a part of oneDPL `2022.6.0` but missed in the release notes until now.

Author: I've removed the line.

and provides high-productivity APIs aimed to minimize programming efforts of C++ developers
creating efficient heterogeneous applications.

New in 2022.9.0
===============

New Features
------------
- Added parallel range algorithms in ``namespace oneapi::dpl::ranges``: ``fill``, ``move``, ``replace``, ``replace_if``,
  ``remove``, ``remove_if``, ``mismatch``, ``minmax_element``, ``min``, ``max``, ``find_first_of``, ``find_end``,
  ``is_sorted_until``. These algorithms operate with C++20 random access ranges.
- Improved performance of set operation algorithms when using device policies: ``set_union``, ``set_difference``,
  ``set_intersection``, ``set_symmetric_difference``.
- Improved performance of ``copy``, ``fill``, ``for_each``, ``replace``, ``reverse``, ``rotate``, ``transform`` and 30+
  other algorithms with device policies on GPUs when using std::reverse_iterator.
- Added ADL-based customization point ``is_onedpl_indirectly_device_accessible`` which can be used to mark iterator
  types as *indirectly device accessible*. Added public trait ``oneapi::dpl::is_directly_device_accessible[_v]`` to
  query if types are indirectly device accessible.

Fixed Issues
------------

> - Eliminated runtime exceptions encountered when compiling code that called ``inclusive_scan``, ``copy_if``,
>   ``partition``, ``unique``, ``reduce_by_segment``, and related algorithms with device policies using the Intel® SYCL
>   compiler (clang++ driver) without specifying an optimization flag.

Reviewer: ```suggestion
- Eliminated runtime exceptions encountered when compiling code that called ``inclusive_scan``, ``copy_if``,
  ``partition``, ``unique``, ``reduce_by_segment``, and related algorithms with device policies using
  the open source oneAPI DPC++ compiler without specifying an optimization flag.
```
Intel® SYCL compiler (clang++ driver) -> open source oneAPI DPC++ compiler. I assume the note means [oneAPI DPC++ compiler](https://github.com/intel/llvm?tab=readme-ov-file#oneapi-dpc-compile)r. I've added "open source" to distinguish it from [Intel® oneAPI DPC++/C++ Compiler](https://www.intel.com/content/www/us/en/developer/tools/oneapi/dpc-compiler.html).

Author: Thank you for the pointer.  I had taken the name from the most recent release of the project -- https://github.com/intel/llvm/releases/tag/v6.1.0 .  It's curious that they don't match.

- ``histogram`` algorithm requires the output value type to be an integral type no larger than four bytes
  when used with a device policy on hardware that does not support 64-bit atomic operations.
- For ``transform_exclusive_scan`` and ``exclusive_scan`` to run in-place (that is, with the same data
  used for both input and destination) and with an execution policy of ``unseq`` or ``par_unseq``, 
  it is required that the provided input and destination iterators are equality comparable.
  Furthermore, the equality comparison of the input and destination iterator must evaluate to true.
  If these conditions are not met, the result of these algorithm calls is undefined.
- Incorrect results may be produced by ``exclusive_scan``, ``inclusive_scan``, ``transform_exclusive_scan``,
  ``transform_inclusive_scan``, ``exclusive_scan_by_segment``, ``inclusive_scan_by_segment``, ``reduce_by_segment``
  with ``unseq`` or ``par_unseq`` policy when compiled by Intel® oneAPI DPC++/C++ Compiler 2024.1 or earlier
  with ``-fiopenmp``, ``-fiopenmp-simd``, ``-qopenmp``, ``-qopenmp-simd`` options on Linux.
  To avoid the issue, pass ``-fopenmp`` or ``-fopenmp-simd`` option instead.

> 

Reviewer: ```suggestion
 -  With libstdc++ version 10, the compilation error *SYCL kernel cannot use exceptions* occurs when calling  ``adjacent_find``, ``is_sorted``, ``is_sorted_until`` with device policies.

```

> 

Reviewer: ```suggestion
- The range-based ``count_if`` may produce incorrect results on Intel® Data Center GPU Max when the driver version
  is "Rolling 2507.12" and newer.

```

and provides high-productivity APIs aimed to minimize programming efforts of C++ developers
creating efficient heterogeneous applications.

New in 2022.9.0
===============

New Features
------------
- Added parallel range algorithms in ``namespace oneapi::dpl::ranges``: ``fill``, ``move``, ``replace``, ``replace_if``,
  ``remove``, ``remove_if``, ``mismatch``, ``minmax_element``, ``min``, ``max``, ``find_first_of``, ``find_end``,
  ``is_sorted_until``. These algorithms operate with C++20 random access ranges.
- Improved performance of set operation algorithms when using device policies: ``set_union``, ``set_difference``,
  ``set_intersection``, ``set_symmetric_difference``.
- Improved performance of ``copy``, ``fill``, ``for_each``, ``replace``, ``reverse``, ``rotate``, ``transform`` and 30+

>   other algorithms with device policies on GPUs when using std::reverse_iterator.

Reviewer 3: ```suggestion
  other algorithms with device policies on GPUs when using ``std::reverse_iterator``.
```

and provides high-productivity APIs aimed to minimize programming efforts of C++ developers
creating efficient heterogeneous applications.

New in 2022.9.0
===============

New Features
------------
- Added parallel range algorithms in ``namespace oneapi::dpl::ranges``: ``fill``, ``move``, ``replace``, ``replace_if``,
  ``remove``, ``remove_if``, ``mismatch``, ``minmax_element``, ``min``, ``max``, ``find_first_of``, ``find_end``,
  ``is_sorted_until``. These algorithms operate with C++20 random access ranges.
- Improved performance of set operation algorithms when using device policies: ``set_union``, ``set_difference``,
  ``set_intersection``, ``set_symmetric_difference``.
- Improved performance of ``copy``, ``fill``, ``for_each``, ``replace``, ``reverse``, ``rotate``, ``transform`` and 30+
  other algorithms with device policies on GPUs when using ``std::reverse_iterator``.

> - Added ADL-based customization point ``is_onedpl_indirectly_device_accessible`` which can be used to mark iterator

Reviewer 4: Please add a comma between **``is_onedpl_indirectly_device_accessible``** and **which**.

Author: Done.

  types as *indirectly device accessible*. Added public trait ``oneapi::dpl::is_directly_device_accessible[_v]`` to
  query if types are indirectly device accessible.

Fixed Issues
------------
- Eliminated runtime exceptions encountered when compiling code that called ``inclusive_scan``, ``copy_if``,
  ``partition``, ``unique``, ``reduce_by_segment``, and related algorithms with device policies using

>   the open source oneAPI DPC++ compiler without specifying an optimization flag.

Reviewer 4: Please capitalize Compiler in oneAPI DPC++ compiler.

Author: Done.

> - Fixed a compilation error in ``reduce_by_segment`` when called with a device policy.

Reviewer 3: It was fixed in https://github.com/uxlfoundation/oneDPL/pull/2206 and resolves an ambiguity between `std::tuple` and our internal tuple leading to differing returns from the same function. It occurred in `2022.7.0` but was discovered by a customer recently, so I wanted to include a mention in the notes.

It could potentially show up in contexts beyond `zip_iterator`. If we want to be more descriptive with the fix, I suggest the following wording to highlight the error:
```suggestion
- Fixed a compilation error in ``reduce_by_segment`` regarding return type deduction when called with a device policy.
```

- For ``transform_exclusive_scan`` and ``exclusive_scan`` to run in-place (that is, with the same data
  used for both input and destination) and with an execution policy of ``unseq`` or ``par_unseq``, 
  it is required that the provided input and destination iterators are equality comparable.
  Furthermore, the equality comparison of the input and destination iterator must evaluate to true.
  If these conditions are not met, the result of these algorithm calls is undefined.
- Incorrect results may be produced by ``exclusive_scan``, ``inclusive_scan``, ``transform_exclusive_scan``,
  ``transform_inclusive_scan``, ``exclusive_scan_by_segment``, ``inclusive_scan_by_segment``, ``reduce_by_segment``
  with ``unseq`` or ``par_unseq`` policy when compiled by Intel® oneAPI DPC++/C++ Compiler 2024.1 or earlier
  with ``-fiopenmp``, ``-fiopenmp-simd``, ``-qopenmp``, ``-qopenmp-simd`` options on Linux.
  To avoid the issue, pass ``-fopenmp`` or ``-fopenmp-simd`` option instead.
- With libstdc++ version 10, the compilation error *SYCL kernel cannot use exceptions* occurs
  when calling the range-based ``adjacent_find``, ``is_sorted`` or ``is_sorted_until`` with device policies.

> - The range-based ``count_if`` may produce incorrect results on Intel® Data Center GPU Max when the driver version

Reviewer 4: Is Intel® Data Center GPU Max the Bridge or Series version? Intel® Data Center GPU Max is incomplete as a product name.

Reviewer: ```suggestion
- The range-based ``count_if`` may produce incorrect results on Intel® Data Center GPU Max Series when the driver version
```

- For ``transform_exclusive_scan`` and ``exclusive_scan`` to run in-place (that is, with the same data
  used for both input and destination) and with an execution policy of ``unseq`` or ``par_unseq``, 
  it is required that the provided input and destination iterators are equality comparable.
  Furthermore, the equality comparison of the input and destination iterator must evaluate to true.
  If these conditions are not met, the result of these algorithm calls is undefined.
- Incorrect results may be produced by ``exclusive_scan``, ``inclusive_scan``, ``transform_exclusive_scan``,
  ``transform_inclusive_scan``, ``exclusive_scan_by_segment``, ``inclusive_scan_by_segment``, ``reduce_by_segment``
  with ``unseq`` or ``par_unseq`` policy when compiled by Intel® oneAPI DPC++/C++ Compiler 2024.1 or earlier
  with ``-fiopenmp``, ``-fiopenmp-simd``, ``-qopenmp``, ``-qopenmp-simd`` options on Linux.
  To avoid the issue, pass ``-fopenmp`` or ``-fopenmp-simd`` option instead.
- With libstdc++ version 10, the compilation error *SYCL kernel cannot use exceptions* occurs
  when calling the range-based ``adjacent_find``, ``is_sorted`` or ``is_sorted_until`` algorithms with device policies,

>   or using ``std::ranges::take_view`` or ``std::ranges::drop_view`` to construct views passed to range-based algorithms

Author: @MikeDvorskiy please confirm the wording here is correct.

Reviewer: I am not Mikhail, but my additional suggestion is to move the part with the range adaptors t into `introduction.rst` alongside other permanent limitations. We cannot fix it.

Author: I've moved it to the introduction.  Thanks.

Reviewer: I've left a suggestion below to improve the note in the library guide.
